python-package/lightgbm/dask.py

```diff
@@ -102,20 +109,30 @@ def _train(client, data, label, params, model_factory, weight=None, **kwargs):
     y : dask array of shape = [n_samples]
         The target values (class labels in classification, real numbers in regression).
     params : dict
-    model_factory : lightgbm.LGBMClassifier or lightgbm.LGBMRegressor class
+    model_factory : lightgbm.LGBMClassifier, lightgbm.LGBMRegressor, or lightgbm.LGBMRanker class
     sample_weight : array-like of shape = [n_samples] or None, optional (default=None)
-            Weights of training data.
+        Weights of training data.
+    group : array-like
+        Group/query data, used for ranking task. sum(group) = n_samples.
     """
-    # Split arrays/dataframes into parts. Arrange parts into tuples to enforce co-locality
+    # Split arrays/dataframes into parts. Arrange parts into dicts to enforce co-locality
     data_parts = _split_to_parts(data, is_matrix=True)
     label_parts = _split_to_parts(label, is_matrix=False)
-    if weight is None:
-        parts = list(map(delayed, zip(data_parts, label_parts)))
-    else:
-        weight_parts = _split_to_parts(weight, is_matrix=False)
-        parts = list(map(delayed, zip(data_parts, label_parts, weight_parts)))
+    parts = [{'X': x, 'y': y} for (x, y) in zip(data_parts, label_parts)]
+
+    # append weight, group vectors to part dicts when needed.
+    if sample_weight is not None:
+        weight_parts = _split_to_parts(sample_weight, is_matrix=False)
+        for i, d in enumerate(parts):
+            parts[i] = {**d, 'weight': weight_parts[i]}
+
+    if group is not None:
+        group_parts = _split_to_parts(group, is_matrix=False)
+        for i, d in enumerate(parts):
+            parts[i] = {**d, 'group': group_parts[i]}
 
     # Start computation in the background
+    parts = list(map(delayed, parts))
```

Comment:
:star_struck: cool idea, I like this! I think it can be in a separate PR though, together with the "put parts into dicts" change.

Reply:
Reverted to tuples in d6e5209